src/Web/HttpApiData/Internal/FormUrlEncoded.hs

```diff
@@ -0,0 +1,163 @@
+module Web.HttpApiData.Internal.FormUrlEncoded where
+
+#if __GLASGOW_HASKELL__ < 710
+import Control.Applicative
+#endif
+
+import           Control.Arrow           (first, left, second)
+import           Control.Monad.State
+import qualified Data.ByteString.Lazy    as BSL
+import qualified Data.Map                as M
+import           Data.Monoid
+import           Data.String.Conversions (cs)
+import qualified Data.Text               as T
+import           GHC.Exts                (IsList (..))
+import           GHC.Generics
+import           Network.URI             (escapeURIString, isUnreserved,
+                                          unEscapeString)
+
+import Web.HttpApiData.Internal.HttpApiData
+
+-- | The contents of a form, not yet url-encoded.
+newtype Form = Form { unForm :: M.Map T.Text T.Text }
+  deriving (Eq, Show, Read, Generic, Monoid)
+
+instance IsList Form where
+  type Item Form = (T.Text, T.Text)
+  fromList = Form . M.fromList
+  toList = M.toList . unForm
+
+-- | A type that can be converted to @application/x-www-form-urlencoded@
+class ToForm a where
+  toForm :: a -> Form
+  default toForm :: (Generic a, GToForm (Rep a)) => a -> Form
+  toForm = genericToForm
+
+instance ToForm [(T.Text, T.Text)] where toForm = Form . M.fromList
+instance ToForm (M.Map T.Text T.Text) where toForm = Form
+instance ToForm Form where toForm = id
+
+genericToForm :: (Generic a, GToForm (Rep a)) => a -> Form
+genericToForm x = evalState (unGenericToFormS . gToForm $ from x) (Nothing, 0)
+
+-- | A datatype for generically generating forms. The state keeps track of the
+-- record name currently being inspected (if any) and the field number.
+newtype GenericToFormS a = GenericToFormS {
+  unGenericToFormS :: State (Maybe T.Text, Int) a
+  } deriving (Functor, Applicative, Generic, Monad, MonadState (Maybe T.Text, Int))
+
+class GToForm (f :: * -> *) where
+  gToForm :: f x -> GenericToFormS Form
+
+instance (GToForm f, GToForm g) => GToForm (f :*: g) where
+  gToForm (a :*: b) = do
+    a' <- gToForm a
+    modify $ second (+ 1)
+    b' <- gToForm b
+    return (a' <> b')
+
+instance (GToForm f, GToForm g) => GToForm (f :+: g) where
+  gToForm (L1 a) = gToForm a
+  gToForm (R1 a) = gToForm a
+
+instance (GToForm f) => GToForm (M1 D x f) where
+  gToForm (M1 a) = gToForm a
+
+instance (GToForm f) => GToForm (M1 C x f) where
+  gToForm (M1 a) = gToForm a
+
+instance {-# OVERLAPPABLE #-} (Selector sel, GToForm f)
+    => GToForm (M1 S sel f) where
+  gToForm (M1 a) = modify (first $ const sel) >> gToForm a
+    where sel = Just . T.pack $ selName (Proxy3 :: Proxy3 sel g p)
+
+instance (ToHttpApiData f) => GToForm (K1 i f) where
+  gToForm (K1 a) = do
+    s <- get
+    return . Form $ M.insert (name s) (toUrlPiece a) $ mempty
+    where name (Nothing, i) = T.pack $ show i
+          name (Just x,  _) = x
+
+-- | A type that can be converted from @application/x-www-form-urlencoded@,
+-- with the possibility of failure.
```

Comment:
Like with `ToForm` this is description is not quite true.
We only convert from `Form` which we first need to be decode from a `ByteString` with `decodeFormUrlEncoded`. Which we also need to bring here from servant.